test_plans/sycl_ext_oneapi_composite_device.asciidoc

```diff
@@ -0,0 +1,132 @@
+:sectnums:
+:xrefstyle: short
+
+= Test plan for sycl_ext_oneapi_composite_device
+
+This is a test plan for the API described in the
+https://github.com/intel/llvm/blob/e94b24718e60a7fa03ca1abbde4f7e37bbd0557d/sycl/doc/extensions/proposed/sycl_ext_oneapi_composite_device.asciidoc[`sycl_ext_oneapi_composite_device`]
+extension specification.
+
+== Testing scope
+
+=== Device coverage
+
+All of the tests described below are performed only on the default device that
+is selected on the CTS command line.
+
+=== Feature test macro
+
+All of the tests should use `#ifdef SYCL_EXT_ONEAPI_COMPOSITE_DEVICE` so they
+can be skipped if the extension is not supported.
+
+== Tests
+
+Note that some of the sections below closely follow the extension specification
+document and describe tests with narrower scope, which are aimed to check
+specific functionality. There are sections below which describe bigger test
+cases which are intended to check how different elements of the extension work
+together.
+
+=== Enumerating composite devices
+
+==== Through the new `get_composite_devices()` API
+
+This test should check that freestanding function `get_composite_devices()`:
+
+- has the correct signature
+- does not throw an exception
+- returns the same result on subsequent calls
+
+==== Through the new `platform::ext_oneapi_get_composite_devices()` API
+
+This test should take a platform the selected device belongs to and check that
+`platform::ext_oneapi_get_composite_devices()` method:
+
+- has the correct signature
+- does not throw an exception
+- returns the same result on subsequent calls
+
+==== Composite devices are not considered as root devices
+
+The test should ensure that no composite devices are returned through the
+`platform::get_device()` method for a platform the selected device belongs to.
+
+=== New device information descriptors
+
+==== `info::device::component_devices`
+
+The test should check that a query for the information descriptor has correct
+return type.
+
+The test should also check the value returned for the descriptor:
+
+- if the selected device is a component device, an empty vector is expected
+- otherwise, returned vector is expected to be non-empty and each its element
+  should be additionally checked:
+  - each component device should have the `ext_oneapi_is_component` aspect
+  - each component device should return the selected device through
+    `info::device::composite_device` query
+
+==== `info::device::composite_device`
+
+The test should check that a query for the information descriptor has correct
+return type.
+
+The test should also check the value returned for the descriptor:
+
+- if the selected device is a component device, then a synchronous `exception`
+  with `errc::invalid` error code is expected
+- otherwise, returned device should have the `ext_oneapi_is_composite` aspect
+
```

Comment:
I think your test for `info::device::component_devices` won't work because the CTS "selected device" will always be some root device, and root devices are never composite devices.  I'd suggest combining the tests for `info::device::component_devices` and `info::device::composite_device` like this:

* If the selected device is not a composite device, `info::device::component_devices` must return an empty list.  (The expectation is that the selected device will never be a composite device.)
* If the selected device is not a component device, `info::device::composite_device` must raise a synchronous `exception` with the `errc::invalid` error code.
* Otherwise:
  * The returned device must have the `ext_oneapi_is_composite` aspect.
  * Calling `info::device::component_devices` must return a list of at least 2 devices.
    * One of these devices must be the selected device.
    * All devices in the list must have the `ext_oneapi_is_component` aspect.

Reply:
What do you think about just extending each test to cover not only the CTS selected device, but also a composite device which is queried from the same platform?

I'm afraid that if we only test the CTS selected device, then we simply won't have coverage for "positive" case (i.e. we don't check `component_devices` query on a composite device at all)

Reply:
> What do you think about just extending each test to cover not only the CTS selected device, but also a composite device which is queried from the same platform?

What do you mean by "each test"?  Do you mean that we would make some fundamental change to the CTS such that every CTS test also tries to run on a composite device?  Or, are you just talking about the tests in this test plan?

Reply:
> > What do you think about just extending each test to cover not only the CTS selected device, but also a composite device which is queried from the same platform?
> 
> What do you mean by "each test"? Do you mean that we would make some fundamental change to the CTS such that every CTS test also tries to run on a composite device? Or, are you just talking about the tests in this test plan?

I mean only this test plan and moreover, only a few test cases (for info queries and aspects), which depend on the device they get and therefore would have more coverage if we launched them on both kinds of devices

Reply:
> Isn't this similar to what I proposed above:

It is, but I think that selecting a wide range of devices first and then performing the same set of checks on every device from that range is more orderly. For example, in the algorithm described by you we never query `composite_device` info on a composite device

Reply:
> Your proposal is similar, except you want to iterate over all composite devices (returned from get_composite_devices), and then run the composite device tests on those devices.

My implementation uses `platform::ext_oneapi_get_composite_devices()` to at least limit the list to the tested platform. However, I agree that until we have changed the device selection/testing strategy globally, let's stick to the current philosophy. db80d6d44b4ccdaa24bf2a52c1fee0274935e0b0 applies your device selection logic, but leaves the general structure in place: all tests are executed for potentially more than one device, but list of the checks is consistent for both and doesn't depend (that much) on which device we are testing, but mostly on what we are checking for